Readme.md
Buffalo
==================
Buffalo is a lightweight [BSON][1] and [Mongo Wire Protocol][2] library for [Node.js][3]. It was built as the underlying
engine for [Mongolian DeadBeef][4].

The motivation is to make a fast and simple parser and serializer for BSON.

[![Build Status](https://secure.travis-ci.org/marcello3d/node-buffalo.png)](http://travis-ci.org/marcello3d/node-buffalo)

Installation
------------
**DISCLAIMER: The API is experimental. I will be adding, removing, and changing the API in the
interest of a solid API. Use at your own risk**

You can either clone the source and install with `npm link`, or install the latest published version from npm with
`npm install buffalo`.

Running Tests
-------------
Run the tests with `npm test`.

API
---
Buffalo exposes two methods:

    exports.parse = function(buffer) { ... }
    exports.serialize = function(object) { ... }

And several types:

    exports.Long // goog.math.Long - http://closure-library.googlecode.com/svn/docs/class_goog_math_Long.html
    exports.ObjectId = function(buffer) // buffer must be a 12-byte Buffer, accessible via the bytes property
    exports.ObjectId = function(string) // string must be a 24-char hex string
    exports.ObjectId = function() // generates an ObjectId
    exports.Timestamp // under construction

The BSON types are mapped as follows:

+ <code>0x01</code> - Floating point - mapped to <code>Number</code>
+ <code>0x02</code> - UTF-8 string - mapped to <code>String</code>
+ <code>0x03</code> - Embedded document - mapped to <code>Object</code>
+ <code>0x04</code> - Array - mapped to <code>Array</code>
+ <code>0x05</code> - Binary data - mapped to Node.js <code>Buffer</code> (with property <code>subtype</code>)
+ <code>0x06</code> - Undefined - mapped to <code>undefined</code>
+ <code>0x07</code> - ObjectId - mapped to <code>exports.ObjectId</code>
+ <code>0x08</code> - Boolean - mapped to <code>true</code> or <code>false</code>
+ <code>0x09</code> - UTC datetime - mapped to <code>Date</code>
+ <code>0x0A</code> - Null value - mapped to <code>null</code>
+ <code>0x0B</code> - Regular expression - mapped to <code>RegExp</code> (Note: only flags g, i, and m are supported)
+ <code>0x0C</code> - DBPointer - currently unmapped
+ <code>0x0D</code> - JavaScript code - mapped to <code>Function</code> or <code>Object</code> with property <code>code</code>
+ <code>0x0E</code> - Symbol - mapped to <code>String</code>
+ <code>0x0F</code> - JavaScript code w/ scope - mapped to <code>Function</code> or <code>Object</code> with properties <code>code</code> and <code>scope</code>
+ <code>0x10</code> - 32-bit Integer - mapped to <code>Number</code>
+ <code>0x11</code> - Timestamp - mapped to <code>exports.Timestamp</code>
+ <code>0x12</code> - 64-bit integer - mapped to <code>exports.Long</code>
+ <code>0xFF</code> - Min key - currently unmapped
+ <code>0x7F</code> - Max key - currently unmapped

Examples
--------

    var BSON = require('buffalo')

    // Parse a Buffer
    var object = BSON.parse(buffer)

    // Serialize an object
    var buffer = BSON.serialize(object)

Contributing
------------
Try it out and send me feedback! Unit tests and documentation are good, too.

License
-------
Buffalo is open source software under the [zlib license][5].

[1]: http://bsonspec.org/#/specification
[2]: http://www.mongodb.org/display/DOCS/Mongo+Wire+Protocol
[3]: http://nodejs.org/
[4]: https://github.com/marcello3d/node-mongolian
[5]: https://github.com/marcello3d/node-buffalo/blob/master/LICENSE


Changelog.md
Buffalo Changelog
-----------------

v0.1.2

+ new ObjectId methods
+ 0.6.x support
+ added Mongo Wire Protocol support
+ more tests
+ using nodeunit instead of vows
+ fixed UTF-8 support (thanks to Marsup)

v0.1.1

+ optimization brings speed up to par with mongodb native and json (parsing, but not serializing)
+ fixed parsing bug in embedded objects followed by other values + new unit test (#1)

v0.1.0

+ initial release

test/bson-test.js
/* Buffalo by Marcello Bastea-Forte - zlib license */

var BSON = require('../buffalo')
var Long = BSON.Long

function C(character) {
    return character.charCodeAt(0)
}

var bsonSample1 = [
                0x16, 0x00, 0x00, 0x00,
                0x02,
                    C('h'), C('e'), C('l'), C('l'), C('o'), 0x00,
                    0x06, 0x00, 0x00, 0x00, C('w'), C('o'), C('r'), C('l'), C('d'), 0x00,
                0x00]

var bsonSample1Expected = {"hello": "world"}

var bsonSample2 = [
                 0x31, 0x00, 0x00, 0x00,
                     0x04, C('B'), C('S'), C('O'), C('N'), 0x00,
                     0x26, 0x00, 0x00, 0x00,
                         0x02, C('0'), 0x00,
                             0x08, 0x00, 0x00, 0x00, C('a'), C('w'), C('e'), C('s'), C('o'), C('m'), C('e'), 0x00,
                         0x01, C('1'), 0x00,
                             0x33, 0x33, 0x33, 0x33, 0x33, 0x33, 0x14, 0x40,
                         0x10, C('2'), 0x00,
                             0xc2, 0x07, 0x00, 0x00,
                     0x00,
                 0x00]

var bsonSample2Expected = {"BSON": ["awesome", 5.05, 1986]}


function testBackAndForth(originalObject) {
    return function(test) {
        var serialized = BSON.serialize(originalObject)
        test.ok(serialized.length)
        var parsed = BSON.parse(serialized)
        test.deepEqual(parsed, originalObject)
        test.done()
    }
}

module.exports = {
    "sample BSON 1 parses to {\"hello\": \"world\"}": function(test) {
        var parsed = BSON.parse(new Buffer(bsonSample1))
        test.deepEqual(parsed, bsonSample1Expected)
        test.done()
    },
    "sample BSON 2 parses to {\"BSON\": [\"awesome\", 5.05, 1986]}": function(test) {
        var parsed = BSON.parse(new Buffer(bsonSample2))
        test.deepEqual(parsed, bsonSample2Expected)
        test.done()
    },
    "javascript object {\"hello\": \"world\"}": function(test) {
        var serialized = BSON.serialize(bsonSample1Expected)
        test.equal(serialized.toString('base64'), 
                   new Buffer(bsonSample1).toString('base64'))
        test.done()
    },
    "javascript object {\"BSON\": [\"awesome\", 5.05, 1986]}": function(test) {
        var serialized = BSON.serialize(bsonSample2Expected)
        test.equal(serialized.toString('base64'), 
                   new Buffer(bsonSample2).toString('base64'))
        test.done()
    },
    "serialize/parse []": testBackAndForth([]),
    "serialize/parse [1,2,3]": testBackAndForth([1,2,3]),
    "serialize/parse ['a','b','c']": testBackAndForth(['a','b','c']),
    "serialize/parse ['è']": testBackAndForth(['à', 'é', 'è', 'ê', 'ñ', 'ù', 'ç', 'ï']),
    "serialize/parse ['short','longer','even longer... Hamburger swine boudin bresaola, tongue meatball biltong rump. Strip steak pig venison tongue. Tenderloin bresaola brisket jowl ham, flank shankle drumstick tongue hamburger swine. Pastrami hamburger boudin beef ribs, sirloin jerky sausage tail bresaola strip steak pork loin corned beef. Turkey jerky jowl sirloin. Chuck pig hamburger fatback. Corned beef beef brisket swine short ribs shoulder.']": testBackAndForth(['short','longer','even longer... Hamburger swine boudin bresaola, tongue meatball biltong rump. Strip steak pig venison tongue. Tenderloin bresaola brisket jowl ham, flank shankle drumstick tongue hamburger swine. Pastrami hamburger boudin beef ribs, sirloin jerky sausage tail bresaola strip steak pork loin corned beef. Turkey jerky jowl sirloin. Chuck pig hamburger fatback. Corned beef beef brisket swine short ribs shoulder.']),
    "serialize/parse [new Date]": testBackAndForth([new Date]),
    "serialize/parse {}": testBackAndForth({}),
    "serialize/parse {'hello':'there'}": testBackAndForth({'hello':'there'}),
    "serialize/parse {level1:{level2:{level3:{level4:{level5:{level6:{level7:'tada!'}}}}}}}": testBackAndForth({level1:{level2:{level3:{level4:{level5:{level6:{level7:'tada!'}}}}}}}),
    "serialize/parse [1.500343]": testBackAndForth([1.500343]),
    "serialize/parse [-1.500343]": testBackAndForth([-1.500343]),
    "serialize/parse [1<<20]": testBackAndForth([1<<20]),
    "serialize/parse [1<<28]": testBackAndForth([1<<28]),
    "serialize/parse [1<<31]": testBackAndForth([1<<31]),
    "serialize/parse [1<<35]": testBackAndForth([1<<35]),
    "serialize/parse [1<<40]": testBackAndForth([1<<40]),
    "serialize/parse [1<<60]": testBackAndForth([1<<60]),
    "serialize/parse [true]": testBackAndForth([true]),
    "serialize/parse [false]": testBackAndForth([false]),
    "serialize/parse [true,false,true,false,0,1]": testBackAndForth([true,false,true,false,0,1]),
    "serialize/parse [null]": testBackAndForth([null]),
    "serialize/parse [undefined]": testBackAndForth([undefined]),
    "serialize/parse [1234567890]": testBackAndForth([1234567890]),
    "serialize/parse [1234567890123456]": testBackAndForth([1234567890123456]),
    "serialize/parse [12345678901234567890]": testBackAndForth([12345678901234567890]),
    "serialize/parse [123456789012345678901234567890]": testBackAndForth([123456789012345678901234567890]),
    "serialize/parse [1234567890123456.78901234567890]": testBackAndForth([1234567890123456.78901234567890]),
    "serialize/parse [new Long(100,100)]": testBackAndForth([new Long(100,100)]),
    "serialize/parse [-1234567890]": testBackAndForth([-1234567890]),
    "serialize/parse [-1234567890123456]": testBackAndForth([-1234567890123456]),
    "serialize/parse [-12345678901234567890]": testBackAndForth([-12345678901234567890]),
    "serialize/parse [-123456789012345678901234567890]": testBackAndForth([-123456789012345678901234567890]),
    "serialize/parse [-1234567890123456.78901234567890]": testBackAndForth([-1234567890123456.78901234567890]),
    "serialize/parse [new Buffer(\"hello\", \"utf8\")]": function(test) {
        var serialized = BSON.serialize([new Buffer("hello", "utf8")])
        test.ok(serialized.length)
        var parsed = BSON.parse(serialized)
        test.ok(parsed[0] instanceof Buffer)
        test.equal(parsed[0].toString(), "hello")
        test.done()
    },
    "serialize/parse complex object": testBackAndForth({
        string: "Strings are great",
        decimal: 3.14159265,
        bool: true,
        integer: 5,
        subObject: {
            moreText: "Bacon ipsum dolor sit amet cow pork belly rump ribeye pastrami andouille. Tail hamburger pork belly, drumstick flank salami t-bone sirloin pork chop ribeye ham chuck pork loin shankle. Ham fatback pork swine, sirloin shankle short loin andouille shank sausage meatloaf drumstick. Pig chicken cow bresaola, pork loin jerky meatball tenderloin brisket strip steak jowl spare ribs. Biltong sirloin pork belly boudin, bacon pastrami rump chicken. Jowl rump fatback, biltong bacon t-bone turkey. Turkey pork loin boudin, tenderloin jerky beef ribs pastrami spare ribs biltong pork chop beef.",
            longKeylongKeylongKeylongKeylongKeylongKey: "Pork belly boudin shoulder ribeye pork chop brisket biltong short ribs. Salami beef pork belly, t-bone sirloin meatloaf tail jowl spare ribs. Sirloin biltong bresaola cow turkey. Biltong fatback meatball, bresaola tail shankle turkey pancetta ham ribeye flank bacon jerky pork chop. Boudin sirloin shoulder, salami swine flank jerky t-bone pork chop pork beef tongue. Bresaola ribeye jerky andouille. Ribeye ground round sausage biltong beef ribs chuck, shank hamburger chicken short ribs spare ribs tenderloin meatloaf pork loin."
        },
        subArray: [1,2,3,4,5,6,7,8,9,10],
        anotherString: "another string"
    }),
    "serialize/parse [/hello/i]": function(test) {
        var serialized = BSON.serialize([/hello/i])
        test.ok(serialized.length)
        var parsed = BSON.parse(serialized)
        test.ok(parsed[0] instanceof RegExp)
        test.equal(parsed[0].toString(), '/hello/i')
        test.done()
    },
    "serialize/parse [/hello[.]+/i]": function(test) {
        var serialized = BSON.serialize([/hello[.]+/i])
        test.ok(serialized.length)
        var parsed = BSON.parse(serialized)
        test.ok(parsed[0] instanceof RegExp)
        test.equal(parsed[0].toString(),"/hello[.]+/i")
        test.done()
    },
    "serialize/parse [function(){}]": function(test) {
        var serialized = BSON.serialize([function(){}])
        test.ok(serialized.length)
        var parsed = BSON.parse(serialized)
        test.equal(typeof parsed[0], 'function')
        // We need a fuzzy match, since serializing jumbles the function slightly
        test.ok(/^function[^\(]*\(\){}/.test(parsed[0].toString().replace(/[\s\n]+/g,'')))
        test.done()
    },
    "serialize/parse [function(x,y){ return x + y }]": function(test) {
        var serialized = BSON.serialize([function(x,y){
            return x + y
        }])
        test.ok(serialized.length)
        var parsed = BSON.parse(serialized)
        test.equal(typeof parsed[0], 'function')
        test.ok(/function[^(]*\(x,y\)\{returnx\+y\}/.test(parsed[0].toString().replace(/[\s\n]+/g,'')))
        test.equal(parsed[0](4,5), 9)
        test.done()
    },
    "serialize/parse scoped function(){ return x + y } (with x=4, y=5)": function(test) {
        var func = function() {
            return x + y
        }
        func.scope = {
            x:4,
            y:5
        }
        var serialized = BSON.serialize([func])
        test.ok(serialized.length)
        var parsed = BSON.parse(serialized)
        test.equal(typeof parsed[0], 'function')
        test.ok(/function[^(]*\(\)\{with\({x:4,y:5}\){returnx\+y\}}/.test(parsed[0].toString().replace(/[\s\n]+/g,'')))
        test.equal(parsed[0](), 9)
        test.done()
    }
}


lib/binary.js
/* Buffalo by Marcello Bastea-Forte - zlib license */

var IEEE754 = require('./extern/buffer_ieee754')

exports.writeCString = function(buffer, offset, string) {
    offset += buffer.write(string, offset, 'utf8')
    buffer[offset++] = 0
    return offset
}
exports.writeString = function(buffer, offset, string) {
    offset = exports.writeInt(buffer, offset, Buffer.byteLength(string) + 1) // include \x00 character in count
    return exports.writeCString(buffer, offset, string)
}
exports.writeInt = function(buffer, offset, num) {
    buffer[offset++] = (num) & 0xff
    buffer[offset++] = (num>>8) & 0xff
    buffer[offset++] = (num>>16) & 0xff
    buffer[offset++] = (num>>24) & 0xff
    return offset
}
exports.writeInt3 = function(buffer, offset, num) {
    buffer[offset++] = (num) & 0xff
    buffer[offset++] = (num>>8) & 0xff
    buffer[offset++] = (num>>16) & 0xff
    return offset
}
exports.writeDouble = function(buffer, offset, num) {
    IEEE754.writeIEEE754(buffer, num, offset, 'little', 52, 8)
    return offset + 8
}
exports.writeLong = function(buffer, offset, num) {
    offset = exports.writeInt(buffer, offset, num.low_)
    return exports.writeInt(buffer, offset, num.high_)
}

lib/objectid.js
/* Buffalo by Marcello Bastea-Forte - zlib license */

var crypto = require('crypto')
var os = require('os')

var binary = require('./binary')

module.exports = ObjectId

// Generate machine hash
//  from docs: This is the first three bytes of the (md5) hash of the machine host name, or of the mac/network address,
//     or the virtual machine id.

var machineHash = crypto.createHash('md5').update(os.hostname()).digest('binary')

// Pre-cache the machine hash / pid
var machineAndPid = [
    machineHash[0],
    machineHash[1],
    machineHash[2],
    (process.pid) & 0xFF,
    (process.pid >> 8) & 0xFF
]

// ObjectId increment
var inc = 1

// 32 bit time
// 24 bit machine id
// 16 bit pid
// 24 bit increment
function ObjectId(bytes) {
    if (Buffer.isBuffer(bytes)) {
        if (bytes.length != 12) throw new Error("Buffer-based ObjectId must be 12 bytes")
        this.bytes = bytes
    } else if (typeof bytes == 'string') {
        if (bytes.length != 24) throw new Error("String-based ObjectId must be 24 bytes")
        if (!/^[0-9a-f]{24}$/i.test(bytes)) throw new Error("String-based ObjectId must in hex-format:" + bytes)
        this.bytes = fromHex(bytes)
    } else if (typeof bytes !== 'undefined') {
        throw new Error("Unrecognized type: "+bytes)
    } else {
        var timestamp = (Date.now() / 1000) & 0xFFFFFFFF
        inc = ~~inc+1 // keep as integer
        this.bytes = new Buffer([
            timestamp>>24,
            timestamp>>16,
            timestamp>>8,
            timestamp,
            machineAndPid[0],
            machineAndPid[1],
            machineAndPid[2],
            machineAndPid[3],
            machineAndPid[4],
            inc>>16,
            inc>>8,
            inc
        ])
    }
}
ObjectId.prototype.toString = function() {
    return toHex(this.bytes)
}

var toHex = function(buffer) {
    return buffer.toString('hex')
}

var fromHex = function(string) {
    return new Buffer(string, 'hex')
}

// This is how we support 0.4.x. Pretty grody, huh?
try {
    if (toHex(new Buffer([0x90])) !='90') throw '0.4.x'
} catch (e) {
    toHex = function(buffer) {
        var s = ''
        for (var i=0; i<buffer.length; i++) {
            s += buffer[i] > 16 ? buffer[i].toString(16) : '0' + buffer[i].toString(16)
        }
        return s
    }
}
try {
    if (fromHex('90')[0] != 0x90) throw '0.4.x'
} catch (e) {
    fromHex = function(string) {
        var buffer = new Buffer(string.length/2)
        for (var i=0; i<string.length; i+=2) {
            buffer[i/2] = parseInt(string.substring(i,i+2),16)
        }
        return buffer
    }
}

lib/bson.js
/* Buffalo by Marcello Bastea-Forte - zlib license */

// based on http://bsonspec.org/#/specification

var Long = require('./extern/long').Long
var IEEE754 = require('./extern/buffer_ieee754')
var toSource = require('tosource')
var binary = require('./binary')
var ObjectId = require('./objectid')
var DBRef = require('./dbref')
var Timestamp = require('./timestamp')

exports.DBRef = DBRef
exports.ObjectId = ObjectId
exports.Long = Long
exports.Timestamp = Timestamp
exports.mongo = require('./mongo')

var FLOAT_TYPE             = 1
var STRING_TYPE            = 2
var EMBEDDED_DOCUMENT_TYPE = 3
var ARRAY_TYPE             = 4
var BINARY_TYPE            = 5
var UNDEFINED_TYPE         = 6 // deprecated
var OBJECT_ID_TYPE         = 7
var BOOLEAN_TYPE           = 8
var DATE_TIME_TYPE         = 9
var NULL_TYPE              = 0x0A
var REG_EXP_TYPE           = 0x0B
var DB_REF_TYPE            = 0x0C // deprecated
var CODE_TYPE              = 0x0D
var SYMBOL_TYPE            = 0x0E
var CODE_WITH_SCOPE_TYPE   = 0x0F
var INT32_TYPE             = 0x10
var TIMESTAMP_TYPE         = 0x11
var INT64_TYPE             = 0x12
var MIN_KEY                = 0xFF
var MAX_KEY                = 0x7F

var BINARY_GENERIC_SUBTYPE      = 0x00
var BINARY_FUNCTION_SUBTYPE     = 0x01
var BINARY_OLD_SUBTYPE          = 0x02
var BINARY_UUID_SUBTYPE         = 0x03
var BINARY_MD5_SUBTYPE          = 0x05
var BINARY_USER_DEFINED_SUBTYPE = 0x80

exports.parse = function(buffer, offset) {
    return parse(buffer, offset || 0, false)
}

exports.serialize = function(object, buffer, offset) {
    buffer = buffer || new Buffer(calculate(object))
    serialize(buffer, offset || 0, object)
    return buffer
}


var calculate = exports.calculate = function(object) {
    var count = 4 // size
    // elements
    if (Array.isArray(object)) {
        for (var i=0; i<object.length; i++) {
            count += calculateElement(String(i), object[i])
        }
    } else {
        for (var key in object) {
            count += calculateElement(key, object[key])
        }
    }
    return count + 1 // zero
}
function calculateElement(name, element) {
    if (element instanceof RegExp) {
        return 1 + // type
            Buffer.byteLength(name, 'utf8') + 1 + // cstring name
            Buffer.byteLength(element.source, 'utf8') + 1 + // cstring regexp source
            (element.global ? 1 : 0) + (element.ignoreCase ? 1 : 0) + (element.multiline ? 1 : 0) + 1 // regexp flags cstring
    }
    switch (typeof element) {
        case 'string':
            return 1 + // type
                Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                4 + Buffer.byteLength(element, 'utf8') + 1 // string

        case 'function':
            if (element.scope) {
                return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                    4 + // size of self + code + scope
                    4 + Buffer.byteLength(element.toString(), 'utf8') + 1 + // string code
                    calculate(element.scope) // scope object
            } else {
                return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 +  // cstring name
                    4 + Buffer.byteLength(element.toString(), 'utf8') + 1 // string code
            }
        case 'number':
            // TODO: handle >32bit longs that aren't doubles?
            if (~~element === element) { // Integer?
                return 1 +  // type
                    Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                    4 // 32bit int
            } else {
                return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                    8 // 64bit float
            }
        case 'undefined':
            return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 // cstring name
        case 'object':
            if (element === null) {
                return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 // cstring name
            } else if (element instanceof Long) {
                return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                    8 // 64bit int
            } else if (element instanceof ObjectId) {
                return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                    element.bytes.length // object id
            } else if (element instanceof Buffer) {
                return 1 +  // type
                    Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                    4 + // buffer length
                    1 + // subtype
                    element.length // buffer contents
            } else if (element instanceof Date) {
                return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                    8 // 64bit int
            } else {
                return 1 + // type
                    Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                    calculate(element) // sub-document
            }
        case 'boolean':
            return 1 + // type
                Buffer.byteLength(name, 'utf8') + 1 + // cstring name
                1 // true/false
    }        
    throw new Error("Unrecognized object type: "+typeof element)
}

function serialize(buffer, offset, object) {
    var start = offset
    offset += 4
    if (Array.isArray(object)) {
        for (var i=0; i<object.length; i++) {
            offset = writeElement(buffer, offset, String(i), object[i])
        }
    } else {
        for (var key in object) {
            offset = writeElement(buffer, offset, key, object[key])
        }
    }

    buffer[offset++] = 0

    binary.writeInt(buffer, start, offset - start)
    
    return offset
}
function writeElement(buffer, offset, name, element) {
    if (element instanceof RegExp) {
        buffer[offset++] = REG_EXP_TYPE
        offset = binary.writeCString(buffer, offset, name)
        offset = binary.writeCString(buffer, offset, element.source)
        offset = binary.writeCString(buffer, offset, (element.global     ? 'g' : '') +
                                              (element.ignoreCase ? 'i' : '') +
                                              (element.multiline  ? 'm' : ''))
    } else switch (typeof element) {
        case 'string':
            buffer[offset++] = STRING_TYPE
            offset = binary.writeCString(buffer, offset, name)
            offset = binary.writeString(buffer, offset, element)
            break
        case 'function':
            if (element.scope) {
                buffer[offset++] = CODE_WITH_SCOPE_TYPE
                offset = binary.writeCString(buffer, offset, name)
                var sizeOffset = offset
                offset += 4
                offset = binary.writeString(buffer, offset, element.toString())
                offset = serialize(buffer, offset, element.scope)
                // go back and fill in the size (size int + code + scope), don't update offset
                binary.writeInt(buffer, sizeOffset, offset - sizeOffset)
            } else {
                buffer[offset++] = CODE_TYPE
                offset = binary.writeCString(buffer, offset, name)
                offset = binary.writeString(buffer, offset, element.toString())
            }
            break
        case 'number':
            // TODO: handle >32bit longs that aren't doubles? Or is that too magical?
            if (~~element === element) { // Integer?
                buffer[offset++] = INT32_TYPE
                offset = binary.writeCString(buffer, offset, name)
                offset = binary.writeInt(buffer, offset, element)
            } else {
                buffer[offset++] = FLOAT_TYPE
                offset = binary.writeCString(buffer, offset, name)
                offset = binary.writeDouble(buffer, offset, element)
            }
            break
        case 'undefined':
            buffer[offset++] = UNDEFINED_TYPE
            offset = binary.writeCString(buffer, offset, name)
            break
        case 'object':
            if (element === null) {
                buffer[offset++] = NULL_TYPE
                offset = binary.writeCString(buffer, offset, name)
            } else if (element instanceof Long) {
                buffer[offset++] = (element instanceof Timestamp) ? TIMESTAMP_TYPE : INT64_TYPE
                offset = binary.writeCString(buffer, offset, name)
                offset = binary.writeInt(buffer, offset, element.low_)
                offset = binary.writeInt(buffer, offset, element.high_)
            } else if (element instanceof ObjectId) {
                buffer[offset++] = OBJECT_ID_TYPE
                offset = binary.writeCString(buffer, offset, name)
                element.bytes.copy(buffer, offset)
                offset += element.bytes.length
            } else if (element instanceof Buffer) {
                buffer[offset++] = BINARY_TYPE
                offset = binary.writeCString(buffer, offset, name)
                offset = binary.writeInt(buffer, offset, element.length) // Just the length of the binary
                buffer[offset++] = element.subtype || BINARY_GENERIC_SUBTYPE
                element.copy(buffer, offset)
                offset += element.length
            } else if (element instanceof Date) {
                buffer[offset++] = DATE_TIME_TYPE
                offset = binary.writeCString(buffer, offset, name)
                offset = binary.writeLong(buffer, offset, Long.fromNumber(element.getTime()))
            } else if (Array.isArray(element)) {
                buffer[offset++] = ARRAY_TYPE
                offset = binary.writeCString(buffer, offset, name)
                offset = serialize(buffer, offset, element)
            } else {
                buffer[offset++] = EMBEDDED_DOCUMENT_TYPE
                offset = binary.writeCString(buffer, offset, name)
                offset = serialize(buffer, offset, element)
            }
            break
        case 'boolean':
            buffer[offset++] = BOOLEAN_TYPE
            offset = binary.writeCString(buffer, offset, name)
            buffer[offset++] = element ? 1 : 0
            break
        default:
            throw new Error("Unrecognized object type: "+typeof element)
    }
    return offset
}



var FUNCTION_MATCH = /^\s*function(?:\s+\S+)?\s*\(([^\)]*)\)\s*\{([\s\S]*)\}\s*$/

function parse(buffer, offset, isArray) {
    var length = readInt()
    
    // Check for too-short buffer
    if (buffer.length - offset + 4 < length) throw new Error("Incomplete BSON buffer (got "+(buffer.length - offset + 4)+" bytes, expected "+length+")")

    var object = isArray ? new Array : new Object
    while (true) {
        var type = buffer[offset++]
        if (type == 0) break
        var name = readCString()
        switch (type) {
            case FLOAT_TYPE:
                object[name] = IEEE754.readIEEE754(buffer, offset, 'little', 52, 8)
                offset += 8
                break
            case INT32_TYPE:
                object[name] = readInt()
                break
            case INT64_TYPE:
                object[name] = new Long(readInt(), readInt())
                break
            case STRING_TYPE:    
            case SYMBOL_TYPE:
                object[name] = readString()
                break
            case EMBEDDED_DOCUMENT_TYPE:
                object[name] = readEmbeddedObject(false)
                break
            case ARRAY_TYPE:
                object[name] = readEmbeddedObject(true)
                break
            case BINARY_TYPE:
                var binaryLength = readInt()
                var subtype = buffer[offset++]
                var binary = buffer.slice(offset, offset += binaryLength)
                binary.subtype = subtype
                object[name] = binary
                break
            case OBJECT_ID_TYPE:
                object[name] = new ObjectId(buffer.slice(offset, offset += 12))
                break
            case BOOLEAN_TYPE:
                object[name] = buffer[offset++] == 1
                break
            case DATE_TIME_TYPE:
                object[name] = new Date(new Long(readInt(), readInt()).toNumber())
                break
            case REG_EXP_TYPE:
                object[name] = new RegExp(readCString(), readCString())
                break
            case DB_REF_TYPE:
                object[name] = new DBRef(readString(), new ObjectId(buffer.slice(offset, offset += 12)))
                break
            case CODE_TYPE:
                object[name] = makeFunction(readString())
                break
            case CODE_WITH_SCOPE_TYPE:
                readInt() // ignoring the size, we don't need it
                object[name] = makeFunction(readString(), readEmbeddedObject())
                break
            case TIMESTAMP_TYPE:
                object[name] = new Timestamp(readInt(), readInt())
                break
            case UNDEFINED_TYPE:
                object[name] = undefined
                break
            case NULL_TYPE:
            case MIN_KEY:
            case MAX_KEY:
                object[name] = null
                break
            default:                
                throw new Error("Unknown BSON type: "+type)
        }    
    }
    
    function readEmbeddedObject(isArray) {
        var embeddedLength = readInt()
        offset -= 4 // rewind int size (since parse will read it again)
        var object = parse(buffer, offset, isArray)
        offset += embeddedLength
        return object
    }

    function readInt() {
        // Little-endian
        return (buffer[offset++]) |
               (buffer[offset++] << 8) |
               (buffer[offset++] << 16) |
               (buffer[offset++] << 24)
    }
    function readCString() {
        for (var i=offset; i<buffer.length; i++) {
            if (buffer[i] === 0) {
                return buffer.toString('utf8', offset, (offset = i + 1) - 1)
            }
        }
        throw new Error("Unterminated c-string!")
    }
    function readString() {
        var length = readInt()
        return buffer.toString('utf8', offset, (offset += length) - 1)
    }
    function makeFunction(code, scope) {
        var f
        try {
            var args = [], funcParts = FUNCTION_MATCH.exec(code)
            if (funcParts) {
                args = funcParts[1].split(',').map(function(name) { return name.trim() })
                code = funcParts[2]
            }
            if (scope) code = "with("+toSource(scope)+"){"+code+"}"
            f = new Function(args, code)
        } catch (e) {
            f = {}
        }
        f.code = code
        if (scope) f.scope = scope
        return f
    }

    return object
}
